src/main/java/gregtech/api/capability/impl/MultiblockRecipeLogic.java

```diff
@@ -130,30 +141,33 @@ protected void trySearchNewRecipeDistinct() {
         // each bus individually instead of the combined inventory all at once.
         for (int i = 0; i < importInventory.size(); i++) {
             IItemHandlerModifiable bus = importInventory.get(i);
-            boolean inputsChanged = hasNotifiedInputs();
-            // If the inputs have changed since the last recipe calculation, find a new recipe based on the new inputs
-            if (inputsChanged) {
-                currentRecipe = findRecipe(maxVoltage, bus, importFluids, MatchingMode.DEFAULT);
-                // Cache the current recipe, if one is found
-                if (currentRecipe != null) {
-                    this.previousRecipe = currentRecipe;
-                }
+            // Skip this bus if no recipe was found last time and the inventory did not change
+            if (invalidatedInputList.contains(bus) && !metaTileEntity.getNotifiedItemInputList().contains(bus)) {
+                continue;
+            } else {
+                invalidatedInputList.remove(bus);
+            }
+            // Look for a new recipe after a cache miss
+            currentRecipe = findRecipe(maxVoltage, bus, importFluids, MatchingMode.DEFAULT);
+            // Cache the current recipe, if one is found
+            if (currentRecipe != null) {
+                this.previousRecipe = currentRecipe;
             }
 
-            invalidInputsForRecipes = (currentRecipe == null);
+            if (currentRecipe == null) {
+                invalidatedInputList.add(bus);
+            }
 
             if (currentRecipe != null && setupAndConsumeRecipeInputs(currentRecipe, importInventory.get(i))) {
                 lastRecipeIndex = i;
                 setupRecipe(currentRecipe);
-                metaTileEntity.getNotifiedItemInputList().remove(i);
-                metaTileEntity.getNotifiedFluidInputList().remove(i);
-                break;
+                metaTileEntity.getNotifiedItemInputList().remove(bus);
+                return;
             }
```

Comment:
Can this just be turned into one statement?

Check if current recipe is null -> cache the recipe -> then check if `setupAndConsumeRecipeInputs` -> save index, setup recipe, and remove bus from notified list. 
Else if recipe is null, invalidate the bus